File: src/common/wasm/zig/heaviside.zig
//! WASM element-wise heaviside step function kernels for numeric types.
//!
//! Scalar: out[i] = x1[i] < 0 ? 0 : x1[i] == 0 ? x2 : 1
//! Binary: out[i] = x1[i] < 0 ? 0 : x1[i] == 0 ? x2[i] : 1
//! Operates on contiguous 1D buffers of length N.

const simd = @import("simd.zig");

/// Heaviside step function for f64 (scalar x2): out[i] = x1[i]<0?0 : x1[i]==0?x2 : 1.
export fn heaviside_scalar_f64(x1: [*]const f64, out: [*]f64, N: u32, x2: f64) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        const v = x1[i];
        out[i] = if (v < 0.0) 0.0 else if (v == 0.0) x2 else 1.0;
    }
}

/// Heaviside step function for f32 (scalar x2): out[i] = x1[i]<0?0 : x1[i]==0?x2 : 1.
export fn heaviside_scalar_f32(x1: [*]const f32, out: [*]f32, N: u32, x2: f32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        const v = x1[i];
        out[i] = if (v < 0.0) 0.0 else if (v == 0.0) x2 else 1.0;
    }
}

/// Heaviside step function for f64 (binary x2): out[i] = x1[i]<0?0 : x1[i]==0?x2[i] : 1.
export fn heaviside_f64(x1: [*]const f64, x2: [*]const f64, out: [*]f64, N: u32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        const v = x1[i];
        out[i] = if (v < 0.0) 0.0 else if (v == 0.0) x2[i] else 1.0;
    }
}

/// Heaviside step function for f32 (binary x2): out[i] = x1[i]<0?0 : x1[i]==0?x2[i] : 1.
export fn heaviside_f32(x1: [*]const f32, x2: [*]const f32, out: [*]f32, N: u32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        const v = x1[i];
        out[i] = if (v < 0.0) 0.0 else if (v == 0.0) x2[i] else 1.0;
    }
}

// --- Tests ---

test "heaviside_scalar_f64 basic" {
    const testing = @import("std").testing;
    const x1 = [_]f64{ -2, -1, 0, 1, 2 };
    var out: [5]f64 = undefined;
    heaviside_scalar_f64(&x1, &out, 5, 0.5);
    try testing.expectApproxEqAbs(out[0], 0.0, 1e-10);
    try testing.expectApproxEqAbs(out[1], 0.0, 1e-10);
    try testing.expectApproxEqAbs(out[2], 0.5, 1e-10);
    try testing.expectApproxEqAbs(out[3], 1.0, 1e-10);
    try testing.expectApproxEqAbs(out[4], 1.0, 1e-10);
}

test "heaviside_f64 binary" {
    const testing = @import("std").testing;
    const x1 = [_]f64{ -1, 0, 0, 1 };
    const x2 = [_]f64{ 99, 0.5, 0.7, 99 };
    var out: [4]f64 = undefined;
    heaviside_f64(&x1, &x2, &out, 4);
    try testing.expectApproxEqAbs(out[0], 0.0, 1e-10);
    try testing.expectApproxEqAbs(out[1], 0.5, 1e-10);
    try testing.expectApproxEqAbs(out[2], 0.7, 1e-10);
    try testing.expectApproxEqAbs(out[3], 1.0, 1e-10);
}

test "heaviside_scalar_f32 basic" {
    const testing = @import("std").testing;
    const x1 = [_]f32{ -2, -1, 0, 1, 2 };
    var out: [5]f32 = undefined;
    heaviside_scalar_f32(&x1, &out, 5, 0.5);
    try testing.expectApproxEqAbs(out[0], 0.0, 1e-5);
    try testing.expectApproxEqAbs(out[1], 0.0, 1e-5);
    try testing.expectApproxEqAbs(out[2], 0.5, 1e-5);
    try testing.expectApproxEqAbs(out[3], 1.0, 1e-5);
    try testing.expectApproxEqAbs(out[4], 1.0, 1e-5);
}

test "heaviside_f32 binary" {
    const testing = @import("std").testing;
    const x1 = [_]f32{ -1, 0, 0, 1 };
    const x2 = [_]f32{ 99, 0.5, 0.7, 99 };
    var out: [4]f32 = undefined;
    heaviside_f32(&x1, &x2, &out, 4);
    try testing.expectApproxEqAbs(out[0], 0.0, 1e-5);
    try testing.expectApproxEqAbs(out[1], 0.5, 1e-5);
    try testing.expectApproxEqAbs(out[2], 0.7, 1e-5);
    try testing.expectApproxEqAbs(out[3], 1.0, 1e-5);
}

test "heaviside_scalar_f64 x2=0" {
    const testing = @import("std").testing;
    const x1 = [_]f64{ -1, 0, 1 };
    var out: [3]f64 = undefined;
    heaviside_scalar_f64(&x1, &out, 3, 0.0);
    try testing.expectApproxEqAbs(out[0], 0.0, 1e-10);
    try testing.expectApproxEqAbs(out[1], 0.0, 1e-10);
    try testing.expectApproxEqAbs(out[2], 1.0, 1e-10);
}

test "heaviside_scalar_f64 x2=1" {
    const testing = @import("std").testing;
    const x1 = [_]f64{ -1, 0, 1 };
    var out: [3]f64 = undefined;
    heaviside_scalar_f64(&x1, &out, 3, 1.0);
    try testing.expectApproxEqAbs(out[0], 0.0, 1e-10);
    try testing.expectApproxEqAbs(out[1], 1.0, 1e-10);
    try testing.expectApproxEqAbs(out[2], 1.0, 1e-10);
}

File: src/common/wasm/heaviside.ts
/**
 * WASM-accelerated element-wise heaviside step function.
 *
 * Scalar: out[i] = x1[i] < 0 ? 0 : x1[i] == 0 ? x2 : 1
 * Binary: out[i] = x1[i] < 0 ? 0 : x1[i] == 0 ? x2[i] : 1
 * Returns null if WASM can't handle this case.
 */

import {
  heaviside_scalar_f64,
  heaviside_scalar_f32,
  heaviside_f64,
  heaviside_f32,
} from './bins/heaviside.wasm';
import { ensureMemory, resetAllocator, copyIn, alloc, copyOut } from './runtime';
import { ArrayStorage } from '../storage';
import type { DType, TypedArray } from '../dtype';
import { wasmConfig } from './config';

const BASE_THRESHOLD = 64;

type ScalarFn = (x1Ptr: number, outPtr: number, N: number, x2: number) => void;
type BinaryFn = (x1Ptr: number, x2Ptr: number, outPtr: number, N: number) => void;

const scalarKernels: Partial<Record<DType, ScalarFn>> = {
  float64: heaviside_scalar_f64,
  float32: heaviside_scalar_f32,
};

const binaryKernels: Partial<Record<DType, BinaryFn>> = {
  float64: heaviside_f64,
  float32: heaviside_f32,
};

type AnyTypedArrayCtor = new (length: number) => TypedArray;
const ctorMap: Partial<Record<DType, AnyTypedArrayCtor>> = {
  float64: Float64Array,
  float32: Float32Array,
};

export function wasmHeavisideScalar(
  x1: ArrayStorage,
  x2: number,
  resultDtype: 'float64' | 'float32'
): ArrayStorage | null {
  if (!x1.isCContiguous) return null;
  const size = x1.size;
  if (size < BASE_THRESHOLD * wasmConfig.thresholdMultiplier) return null;

  const kernel = scalarKernels[resultDtype];
  const Ctor = ctorMap[resultDtype];
  if (!kernel || !Ctor) return null;

  const bpe = (Ctor as unknown as { BYTES_PER_ELEMENT: number }).BYTES_PER_ELEMENT;
  ensureMemory(size * bpe * 2);
  resetAllocator();

  const x1Ptr = copyIn(x1.data.subarray(x1.offset, x1.offset + size) as TypedArray);
  const outPtr = alloc(size * bpe);
  kernel(x1Ptr, outPtr, size, x2);

  const outData = copyOut(
    outPtr,
    size,
    Ctor as unknown as new (buf: ArrayBuffer, off: number, len: number) => TypedArray
  );
  return ArrayStorage.fromData(outData, Array.from(x1.shape), resultDtype);
}

export function wasmHeaviside(
  x1: ArrayStorage,
  x2: ArrayStorage,
  resultDtype: 'float64' | 'float32'
): ArrayStorage | null {
  if (!x1.isCContiguous || !x2.isCContiguous) return null;
  const size = x1.size;
  if (size < BASE_THRESHOLD * wasmConfig.thresholdMultiplier) return null;

  const kernel = binaryKernels[resultDtype];
  const Ctor = ctorMap[resultDtype];
  if (!kernel || !Ctor) return null;

  const bpe = (Ctor as unknown as { BYTES_PER_ELEMENT: number }).BYTES_PER_ELEMENT;
  ensureMemory(size * bpe * 3);
  resetAllocator();

  const x1Ptr = copyIn(x1.data.subarray(x1.offset, x1.offset + size) as TypedArray);
  const x2Ptr = copyIn(x2.data.subarray(x2.offset, x2.offset + size) as TypedArray);
  const outPtr = alloc(size * bpe);
  kernel(x1Ptr, x2Ptr, outPtr, size);

  const outData = copyOut(
    outPtr,
    size,
    Ctor as unknown as new (buf: ArrayBuffer, off: number, len: number) => TypedArray
  );
  return ArrayStorage.fromData(outData, Array.from(x1.shape), resultDtype);
}
